### PR TITLE
Add support for non-equi outer joins (v1)

### PR DESCRIPTION
**Pull request by losipiuk**
Add support for non-equi conjuncts in join condition for outer joins.
Supported case is when conjunct is resolvable as a whole using symbols
of inner join table. In such case conjunct is pushed down as filter to
inner source relation.

### CODE REVIEW

**Comment by losipiuk** [comment]
This is another extension of support for non-equi joins condition.
This time for outer joins where conjunct can be resolved using only inner relation.

@cberner, @martint please take a look.

**Comment by losipiuk** [comment]
fixes: https://github.com/facebook/presto/issues/3889

**Comment by losipiuk** [comment]
Last commit is buggy - please do not review it yet - I will let you know when I fix the problem.

**Comment by losipiuk** [comment]
@pnowojski please take a look before I pass it to FB guys.

File: presto-main/src/main/java/com/facebook/presto/operator/InMemoryJoinHash.java [hunk 1]
```diff
@@ -103,19 +105,28 @@ public long getInMemorySizeInBytes()
...
     @Override
-    public long getJoinPosition(int position, Page page, int rawHash)
+    public long getJoinPosition(int rightPosition, Page hashChannelsPage, Page allChannelsPage, int rawHash)
     {
         int pos = getHashPosition(rawHash, mask);
 
         while (key[pos] != -1) {
-            if (positionEqualsCurrentRow(key[pos], position, page.getBlocks())) {
-                return key[pos];
+            if (positionEqualsCurrentRow(key[pos], rightPosition, hashChannelsPage.getBlocks(), allChannelsPage.getBlocks())) {
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/InMemoryJoinHash.java]: compare performance using jmh benchmarks for joining

File: presto-main/src/main/java/com/facebook/presto/operator/InMemoryJoinHash.java [hunk 2]
```diff
@@ -103,19 +105,28 @@ public long getInMemorySizeInBytes()
...
+                long candidate = key[pos];
+                if (filterFunctionPresent) {
+                    while (candidate != -1 && !checkFilterFunctionCurrentRow(Ints.checkedCast(candidate), rightPosition, allChannelsPage.getBlocks())) {
+                        candidate = getNextJoinPosition(candidate, rightPosition, allChannelsPage);
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/InMemoryJoinHash.java]: extract common code to something like `getFirstValidPositionAfter(...)` as we talked.

File: presto-main/src/main/java/com/facebook/presto/operator/InMemoryJoinHash.java [hunk 3]
```diff
@@ -153,13 +174,22 @@ private int hashPosition(int position)
         return pagesHashStrategy.hashPosition(blockIndex, blockPosition);
     }
 
-    private boolean positionEqualsCurrentRow(int leftPosition, int rightPosition, Block... rightBlocks)
+    private boolean positionEqualsCurrentRow(int leftPosition, int rightPosition, Block[] rightHashBlocks, Block[] rightBlocks)
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/InMemoryJoinHash.java]: why this method can not handle filter function? In other words, why can't you push down filter function to `pagesHashStrategy`?

File: presto-main/src/main/java/com/facebook/presto/operator/InMemoryJoinHash.java [hunk 4]
```diff
@@ -153,13 +174,22 @@ private int hashPosition(int position)
...
+    {
+        long pageAddress = addresses.getLong(leftPosition);
+        int blockIndex = decodeSliceIndex(pageAddress);
+        int blockPosition = decodePosition(pageAddress);
+
+        return pagesHashStrategy.positionEqualsRow(blockIndex, blockPosition, rightPosition, rightHashBlocks);
+    }
+
+    private boolean checkFilterFunctionCurrentRow(int leftPosition, int rightPosition, Block[] rightBlocks)
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/InMemoryJoinHash.java]: `check` suggests that it throws an exception. Rename it

File: presto-main/src/main/java/com/facebook/presto/operator/PagesHashStrategy.java [hunk 5]
```diff
@@ -72,4 +75,14 @@
...
+
+    /**
+     * Returns filter function assigned to this PagesHashStrategy.
+     */
+    Optional<JoinFilterFunction> getFilterFunction();
+
+    /**
+     * Checks result of filter function for given row.
+     */
+    boolean checkFilterFunction(int leftBlockIndex, int leftPosition, int rightPosition, Block[] allRightBlocks);
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/PagesHashStrategy.java]: rename `check`

**Comment by pnowojski** [comment]
add unit tests in `TestHashJoinOperator`

**Comment by losipiuk** [comment]
supeseded by v2 (#4950)

**Pull request closed without merging**